File: App.Application.CSharp/Shared/Events/DomainEventFactory.cs
using App.Domain.Shared;
using App.Domain.Shared.EventHelpers;
using App.Domain.Time;

namespace App.Application.CSharp.Shared.Events;

public static class DomainEventFactory
{
    public static DomainEvent<T> Create<T>(
        ushort schemaVer,
        IClock clock,
        IGuid guid,
        Guid correlationId,
        Guid? causationId,
        T payload
    )
    {
        var header = new EventHeader(guid.NewGuid(), schemaVer, clock.UtcNow, correlationId, causationId);
        return new DomainEvent<T>(header, payload);
    }
}

File: App.Domain/Shared/DomainEventFactory.fs
module App.Domain.Shared.Utils.DomainEventFactory

open System
open App.Domain.Shared
open App.Domain.Shared.EventHelpers
open App.Domain.Time

let create<'T>
    (aggregateVer: uint)
    (schemaVer: uint16)
    (occuredAt: DateTimeOffset)
    (id: Guid)
    (correlationId: Guid)
    (causationId: Guid)
    (payload: 'T)
    : DomainEvent<'T> =

    let header =
        { EventId = id
          AggregateVersion = aggregateVer
          SchemaVer = schemaVer
          OccurredAt = occuredAt
          CorrelationId = correlationId
          CausationId = causationId }
        : EventHeader

    { Header = header; Payload = payload }
